crates/iceberg/src/transform/truncate.rs

```diff
@@ -0,0 +1,177 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+use std::sync::Arc;
+
+use arrow_array::ArrayRef;
+use arrow_schema::DataType;
+
+use crate::Error;
+
+use super::TransformFunction;
+
+pub struct Truncate {
+    width: u32,
+}
+
+impl Truncate {
+    pub fn new(width: u32) -> Self {
+        Self { width }
+    }
+}
+
+impl TransformFunction for Truncate {
+    fn transform(&self, input: ArrayRef) -> crate::Result<ArrayRef> {
+        match input.data_type() {
+            DataType::Int32 => {
+                let width: i32 = self.width.try_into().map_err(|_| {
+                    Error::new(
+                        crate::ErrorKind::DataInvalid,
+                        "width is failed to convert to i32 when truncate Int32Array",
+                    )
+                })?;
+                let res: arrow_array::Int32Array = input
+                    .as_any()
+                    .downcast_ref::<arrow_array::Int32Array>()
+                    .unwrap()
+                    .unary(|v| v - (((v % width) + width) % width));
```

Comment:
We could use `v - v.rem_euclid(width)` instead of `v - (((v % width) + width) % width)`. But it's also fine as it is.

Reply:
Thanks! `v.rem_euclid` is more clear!